eventlogger/src/main/java/gov/cms/ab2d/eventlogger/eventloggers/kinesis/KinesisEventLogger.java

```diff
@@ -1,63 +1,54 @@
 package gov.cms.ab2d.eventlogger.eventloggers.kinesis;
 
-import com.amazonaws.ResponseMetadata;
-import com.amazonaws.services.kinesisfirehose.model.PutRecordRequest;
-import com.amazonaws.services.kinesisfirehose.model.PutRecordResult;
-import com.amazonaws.services.kinesisfirehose.model.Record;
 import com.amazonaws.services.kinesisfirehose.AmazonKinesisFirehose;
-import com.fasterxml.jackson.core.JsonProcessingException;
-import com.fasterxml.jackson.databind.ObjectMapper;
-import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
-import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
-import com.fasterxml.jackson.module.paramnames.ParameterNamesModule;
 import gov.cms.ab2d.eventlogger.EventLogger;
 import gov.cms.ab2d.eventlogger.LoggableEvent;
 import lombok.extern.slf4j.Slf4j;
-import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.beans.factory.annotation.Value;
+import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor;
 import org.springframework.stereotype.Service;
 
-import java.nio.ByteBuffer;
+import java.util.concurrent.Future;
 
 @Service
 @Slf4j
 public class KinesisEventLogger implements EventLogger {
     @Value("${execution.env:dev}")
     private String appEnv;
-
-    @Autowired
-    private AmazonKinesisFirehose client;
-
     @Value("${eventlogger.kinesis.stream.prefix:bfd-insights-ab2d-}")
     private String streamId;
 
+    private final KinesisConfig config;
+    private final AmazonKinesisFirehose client;
+
+    public KinesisEventLogger(KinesisConfig config, AmazonKinesisFirehose client) {
+        this.config = config;
+        this.client = client;
+    }
+
     @Override
     public void log(LoggableEvent event) {
-        String json = null;
+        log(event, false);
+    }
+
+    public void log(LoggableEvent event, boolean block) {
         event.setEnvironment(appEnv);
-        try {
-            json = getJsonString(event) + "\n";
-            Record record = new Record()
-                    .withData(ByteBuffer.wrap(json.getBytes()));
-
-            PutRecordRequest putRecordRequest = new PutRecordRequest();
-            String className = event.getClass().getSimpleName().toLowerCase();
-            putRecordRequest.setDeliveryStreamName(streamId + className);
-            putRecordRequest.setRecord(record);
-
-            PutRecordResult putRecordResult = client.putRecord(putRecordRequest);
-            ResponseMetadata data = putRecordResult.getSdkResponseMetadata();
-            event.setAwsId(data.getRequestId());
-        } catch (Exception ex) {
-            log.error("Error in pushing event to Kinesis " + json + " - " + ex.getMessage());
+        ThreadPoolTaskExecutor ex = config.kinesisLogProcessingPool();
+        KinesisEventProcessor processor = new KinesisEventProcessor(event, client, streamId);
+        Future<Void> future = ex.submit(processor);
+        if (block) {
```

Comment:
Curious what the use case would be when we would want to block?

Reply:
There are two types of logging. SQL and Kinesis. There is only one situation that we log to Kinesis and not SQL - when we're logging every bene search. We don't want that in the database. It's also a flood of events which is why we want the multi threading. In all other cases, the events are more moderate (job changes status for example) and you want one in the DB and one put to Kinesis. You want it blocking if you want the DB table to include the AWS ID. You only get the AWS ID once the thread is complete (and the put was successful). Blocking allows you to wait for that thread to end so you can get it.